optimization/eda/probability-learning.metta

```diff
@@ -0,0 +1,291 @@
+;; - Count knob value occurrences
+;; - Compute probabilities with Laplace smoothing
+;; - Normalize probability distributions
+;;
+;; Data Structures:
+;; - KnobProbabilityModel: mkKnobProbabilityModel (Map KnobId (Map KnobValue Number))
+;; - KnobValue: mkKnobValue Number
+;; - KnobId: NodeId
+
+;; ================================================================================
+;; Type Definitions
+;; ================================================================================
+
+;; Knob Value and Probability Model
+(: KnobValue Type)
+(: mkKnobValue (-> Number KnobValue))
+(: KnobId Type)
+(: KnobProbabilityModel Type)
+(: mkKnobProbabilityModel (-> (Map KnobId (Map KnobValue Number)) (KnobProbabilityModel)))
+
+;; ================================================================================
+;; Probability Learning Policy
+;; ================================================================================
+
+;; Learn probability distributions from selected instances
+; (: learn-probability-model (-> Deme (InstanceSet $score) (KnobProbabilityModel)))
+(= (learn-probability-model $deme $selectedInstances)
+  (let* (
+        (() (println! (learn-probability-model: starting)))
+        ((mkDeme $rep (mkSInstSet $instSet) $id) $deme)
+        (() (println! (learn-probability-model: got rep $rep)))
+        ($discMap (getDiscMap $rep))
+        (() (println! (learn-probability-model: got discMap $discMap)))
+        ($probModel (learn-knob-probabilities $discMap $selectedInstances))
+        (() (println! (learn-probability-model: got probModel $probModel)))
+      )
+      $probModel))
+
+;; Learn probability distributions for each knob from selected instances
+; (: learn-knob-probabilities (-> (MultiMap (DiscSpec $knob)) (InstanceSet $score) (KnobProbabilityModel)))
+(= (learn-knob-probabilities $discMap $selectedInstances)
+  (let* (
+        (() (println! (learn-knob-probabilities: starting)))
+        (() (println! (learn-knob-probabilities: got discMap $discMap)))
+        (() (println! (learn-knob-probabilities: got selectedInstances $selectedInstances)))
+        ($knobCounts (count-knob-values $discMap $selectedInstances))
+        (() (println! (learn-knob-probabilities: got knobCounts $knobCounts)))
+        ($probabilities (compute-probabilities $knobCounts))
+        (() (println! (learn-knob-probabilities: got probabilities $probabilities)))
+      )
+      (mkKnobProbabilityModel $probabilities)))
+
+;; Count occurrences of each knob value in selected instances
+; (: count-knob-values (-> (MultiMap (DiscSpec $knob)) (InstanceSet $score) (Map KnobId (Map KnobValue Number))))
+(= (count-knob-values $discMap $selectedInstances)
+  (let* (
+        (() (println! (count-knob-values: starting)))
+        (() (println! (count-knob-values: got discMap $discMap)))
+        (() (println! (count-knob-values: got selectedInstances $selectedInstances)))
+        ($dkmWrapped (crtDiscKnobMap (mkDscMp $discMap) (mkDscKbMp NilMap) 0))
```

Comment:
This is essentially calculating the initial `discMap` again right? Why not pass it down in the first place?